### DISCUSSION TITLE
Copying things out of a Docker based build

### DISCUSSION THREAD

**Discussion by kaihendry** [community/community#27092] [Actions]
<p>I have a Dockerfile that produces some build artifacts.</p>
<p>I’m struggling to find an example or documentation in this case where I need to copy files out of the build Dockerfile.</p>
<p>Does someone have an example I could please refer to?</p>
<p>Thank you in advance!</p>

**Comment by Simran-B** [comment, marked as answer]
<p>With Docker in general, the following ways to get files out of a container come to mind:</p>
<ul>
<li>
<p>Run the <code>docker cp</code> command</p>
<blockquote>
<p>Copy files/folders between a container and the local filesystem</p>
</blockquote>
<p>This would need to be run on the host system, however.</p>
</li>
<li>
<p>Mount a folder of the host system into the container as a volume, then copy files to the mounted folder inside the container.</p>
</li>
<li>
<p>Use a multi-stage Docker image to only leave the files you are interested in in the container at the end of the image building process. If you want to copy out files when running this image later, then this option is irrelevant.</p>
<pre><code class="lang-auto">FROM &lt;base-image&gt; AS build
...
FROM scratch AS artifacts
COPY --from=build /path/to/file /
</code></pre>
<p>Afterwards, you can save the image as archive with <code>docker save -o artifacts.tar &lt;image-name&gt;</code> and unpack it to get at the contents. Note that there’s a bit more unpacking to do, because Docker saves metadata and all the layers individually (here it is just one layer because of the scratch image, however).</p>
</li>
<li>
<p>Store the build artifacts externally, e.g. by uploading them to a cloud storage provider.</p>
</li>
</ul>
<p>Do you use your Docker image via <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontainer" rel="noopener nofollow ugc"><code>jobs.&lt;job_id&gt;.container.image</code></a> or do you <code>run</code> it like <code>docker run ...</code> in a workflow step?</p>
<p>In the former case, use the <code>volume</code> key to mount a folder into the container:<br>
<a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontainervolumes" rel="noopener nofollow ugc"><code>jobs.&lt;job_id&gt;.container.volumes</code></a></p>
<p>In the latter case, you would also need to do that, but by manually mounting a volume with <code>-v</code> in the <code>docker run</code> command.</p>

**Comment by maxisam** [comment]
There is a better way now  
https://stackoverflow.com/questions/57339971/docker-copy-file-out-of-container-while-building-it
